man/nmtui.1.in
.\" nmtui (1) manual page
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public Licence along
.\" with this manual; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
.\"
.\" Copyright 2014 Red Hat, Inc.
.\"
.TH NMTUI "1" "2014-11-06" "NetworkManager 1.2"

.SH NAME
nmtui \- Text User Interface for controlling NetworkManager
.SH SYNOPSIS
.B nmtui
.RI " [ " edit " | " connect " | " hostname " ] [ ... ] "
.P
.B nmtui\-edit
.RI " [ " connection-id " | " connection-name " ] "
.P
.B nmtui\-connect
.RI " [ " connection-name " | " connection-uuid " | " device-name " | " Wi-Fi-SSID " ] "
.P
.B nmtui\-hostname

.SH DESCRIPTION
.B nmtui
is a curses\(hybased TUI application for interacting with \fINetworkManager\fP.
.P
When starting \fInmtui\fP, the user is prompted to choose the activity to perform
unless it was specified as the first argument.
.P
The supported activities are:
.IP \(em 4
\fIedit\fP: show a connection editor that supports adding, modifying, viewing and deleting
connections. It provides similar functionality as \fInm\-connection\-editor\fP.
.IP \(em 4
\fIconnect\fP: show a list of available connections, with the option to activate or deactivate
them. It provides similar functionality as \fInm\-applet\fP.
.IP \(em 4
\fIhostname\fP: set the system hostname.

.P
Corresponding to above activities, \fInmtui\fP also comes with binaries named
\fInmtui\-edit\fP, \fInmtui\-connect\fP, and \fInmtui-hostname\fP to skip the selection
of the activities.

.SH SEE ALSO
.BR NetworkManager(8),
.BR nmcli(1).
.BR nm\-applet(1).
.BR nm\-connection\-editor(1).


NEWS
=============================================
NetworkManager-1.22
Overview of changes since NetworkManager-1.20
=============================================

This is a snapshot of NetworkManager 1.22 development series.
The API is subject to change and not guaranteed to be compatible
with the later release.
USE AT YOUR OWN RISK.  NOT RECOMMENDED FOR PRODUCTION USE!

* Drop support for BlueZ 4. BlueZ 5 was released in 2012 and should
  nowadays be available everywhere.
* core: declare "startup complete" when device reaches "connected"
  state, even if IP addressing methods did not yet fully complete.
  This changes behavior for unblocking "NetworkManager-wait-online.service",
  and "network-online.target" earlier. If this causes issues in your
  setup, you may need to configure "ipv4.may-fail=no" or "ipv6.may-fail=no",
  which delays reaching "connected" state for the address family accordingly.
* Various bug fixes and improvements.
* libnm: hide NMObject and NMClient typedefs from header files. This prevents
  the user from subclassing these types and is an ABI change (in case somebody
  was doing so).
* libnm: retire deprecated WiMAX API NMDeviceWimax and NMWimaxNsp.
  WiMAX support was removed from NetworkManager in version 1.2 (2016) and no such
  type instances would have been created by NMClient for a while now.
* DHCP: switch "internal" DHCPv4 plugin from code based on systemd to use nettools'
  n-dhcp4 library.
* libnm: heavily internal rework NMClient. This slims down libnm and makes the
  implementation more efficient. NMClient should work now well with a separate
  GMainContext.

=============================================
NetworkManager-1.20
Overview of changes since NetworkManager-1.18
=============================================

This is a new stable release of NetworkManager.  Notable changes include:

* The libnm-glib library, deprecated in favor of libnm since NetworkManager 1.0,
  and disabled by default since NetworkManager 1.12, has now been removed.
* The DHCP client now defaults to "internal". The default can be overriden
  at build time by using the --with-config-dhcp-default option of the
  configure script or at run time by setting the main.dhcp option in the
  configuration file.
* Added support for configuring fq_codel line discipline and mirred action.
* Added a possibility for distributions to ship dispatcher scripts in /usr/lib.
* Drop deprecated setting "main.monitor-connection-files" in NetworkManager.conf.
  This setting now has no more effect and was disabled by default for a long time.
  Instead, after changes, load files explicitly with `nmcli connection load` or
  `nmcli connection reload`.
* Rework parsing team JSON config in libnm and stricter validate settings.
  With this, NetworkManager rejects settings that it considers invalid while
  still allowing setting arbitrary JSON config directly.
* Drop ibft settings plugin. This functionality is now covered by using
  nm-initrd-generator from initrd to pre-generate in-memory profiles.
* Support "suppress_prefixlength" attribute for policy routing rules.
  This is what wg-quick uses for the "Improved Rule-based Routing" solution,
  and the user can now manually configure such policy routing rules.
* Support "wireguard.ip4-auto-default-route" and "wireguard.ip6-auto-default-route".
  This automatically implements the "Improved Rule-based Routing" of wg-quick
  to help avoiding routing loops when setting the default-route on the WireGuard
  interface. Note that this is now enabled by default, so there is a change in behavior
  if your WireGuard connection profiles from before had a default-route (/0)
  in allowed-ips.
* Rework implementation of settings plugins and how profiles are presisted
  to disk. This is a large internal refactoring of the settings plugins that
  allows to migrate a connection profile between plugins.
* In-memory profiles are now only handled by keyfile plugin and will also be
  persisted to /run directory. This allows to restart NetworkManager without
  loosing these profiles and it provides a file-system based API for creating
  in-memory profiles.
* Keyfile plugin now supports a read-only directory of profiles under directory
  "/usr/lib/NetworkManager/system-connections". Such profiles still can be modified
  and deleted via D-Bus, which results in writing profiles to /etc or /run that
  shadow the read-only files.
* Add new D-Bus method AddConnection2() that allows to block autoconnect of
  the profile at the moment when creating the profile. Also add support for
  this API to libnm.
* Add flag "no-reapply" to Update2() D-Bus method. Normally, when a connection
  profile gets modified, this only changes the profile itself. When the profile
  is currently activated on a device, then the device's configuration does not update
  before the profile is fully re-activated or Reapply on the device is called. There
  is an exception to this: the "connection.zone" and the "connection.metered"
  properties take effect immediately. The "no-reapply" flag allows suppressing to
  reapply any properties, so that no changes take effect automatically. The purpose
  is to really only modify the profile itself without changes to the runtime configuration
  of the device.
* Add "ipv6.method=disabled" to disable IPv6 on a device, like also possible for
  IPv4. Until now, the users could only set "ipv6.method=ignore" which means the
  users are free to set IPv6 related sysctl values themselves.
* Added support for Wi-Fi Mesh network.

The following changes were backported to 1.18.x releases between 1.18.0
and 1.18.2 are also present in NetworkManager-1.18:

- 1.18.2:
* Add compatibility with out-of-tree WireGuard module on 5.2 kernels
* Fix parsing of BOOTIF= variables in initrd.
* Accept numeric IPv4 prefix in place of a mask when parsing a command line
  in initrd.
* Don't check connectivity of unconfigured devices.
* Fix PKCS#12 handling in the ifcfg-rh plugin.

clients/tui/nmtui.c
// SPDX-License-Identifier: GPL-2.0+
/*
 * Copyright (C) 2013 Red Hat, Inc.
 */

/**
 * SECTION:nmtui
 * @short_description: nmtui toplevel
 *
 * The top level of nmtui. Exists mostly just to call nmtui_connect(),
 * nmtui_edit(), and nmtui_hostname().
 */

#include "nm-default.h"

#include "nmtui.h"

#include <locale.h>
#include <stdlib.h>

#include "nmt-newt.h"
#include "nm-editor-bindings.h"

#include "nmtui-edit.h"
#include "nmtui-connect.h"
#include "nmtui-hostname.h"

NMClient *nm_client;
static GMainLoop *loop;

typedef NmtNewtForm * (*NmtuiSubprogram) (gboolean is_top, int argc, char **argv);

static const struct {
	const char *name, *shortcut, *arg;
	const char *display_name;
	NmtuiSubprogram func;
} subprograms[] = {
	{ "edit",     "nmtui-edit",     N_("connection"),
	  N_("Edit a connection"),
	  nmtui_edit },
	{ "connect",  "nmtui-connect",  N_("connection"),
	  N_("Activate a connection"),
	  nmtui_connect },
	{ "hostname", "nmtui-hostname", N_("new hostname"),
	  N_("Set system hostname"),
	  nmtui_hostname }
};
static const int num_subprograms = G_N_ELEMENTS (subprograms);
static NmtNewtForm *toplevel_form;

static NmtNewtForm *
quit_func (int argc, char **argv)
{
	if (toplevel_form)
		nmt_newt_form_quit (toplevel_form);

	nmtui_quit ();

	return NULL;
}

static void
main_list_activated (NmtNewtWidget *widget, NmtNewtListbox *listbox)
{
	NmtNewtForm *form;
	NmtuiSubprogram sub;

	sub = nmt_newt_listbox_get_active_key (listbox);
	if (sub) {
		form = sub (FALSE, 0, NULL);
		if (form) {
			nmt_newt_form_show (form);
			g_object_unref (form);
		}
	}
}

static NmtNewtForm *
nmtui_main (gboolean is_top, int argc, char **argv)
{
	NmtNewtForm *form;
	NmtNewtWidget *widget, *ok;
	NmtNewtGrid *grid;
	NmtNewtListbox *listbox;
	NmtNewtButtonBox *bbox;
	int i;

	form = g_object_new (NMT_TYPE_NEWT_FORM,
	                     "title", _("NetworkManager TUI"),
	                     "escape-exits", TRUE,
	                     NULL);

	widget = nmt_newt_grid_new ();
	nmt_newt_form_set_content (form, widget);
	grid = NMT_NEWT_GRID (widget);

	widget = nmt_newt_label_new (_("Please select an option"));
	nmt_newt_grid_add (grid, widget, 0, 0);

	widget = g_object_new (NMT_TYPE_NEWT_LISTBOX,
	                       "height", num_subprograms + 2,
	                       "skip-null-keys", TRUE,
	                       NULL);
	nmt_newt_grid_add (grid, widget, 0, 1);
	nmt_newt_widget_set_padding (widget, 0, 1, 0, 1);
	listbox = NMT_NEWT_LISTBOX (widget);
	g_signal_connect (widget, "activated",
	                  G_CALLBACK (main_list_activated), listbox);

	for (i = 0; i < num_subprograms; i++) {
		nmt_newt_listbox_append (listbox, _(subprograms[i].display_name),
		                         subprograms[i].func);
	}
	nmt_newt_listbox_append (listbox, "", NULL);
	nmt_newt_listbox_append (listbox, _("Quit"), quit_func);

	widget = nmt_newt_button_box_new (NMT_NEWT_BUTTON_BOX_HORIZONTAL);
	nmt_newt_grid_add (grid, widget, 0, 2);
	bbox = NMT_NEWT_BUTTON_BOX (widget);

	ok = nmt_newt_button_box_add_end (bbox, _("OK"));
	g_signal_connect (ok, "activated",
	                  G_CALLBACK (main_list_activated), listbox);

	toplevel_form = form;

	return form;
}

/**
 * nmtui_quit:
 *
 * Causes nmtui to exit.
 */
void
nmtui_quit (void)
{
	g_main_loop_quit (loop);
}

static void
usage (void)
{
	const char *argv0 = g_get_prgname ();
	const char *usage_str = _("Usage");
	int i;

	for (i = 0; i < num_subprograms; i++) {
		if (!strcmp (argv0, subprograms[i].shortcut)) {
			g_printerr ("%s: %s [%s]\n", usage_str, argv0, _(subprograms[i].arg));
			exit (1);
		}
	}

	g_printerr ("%s: nmtui\n", usage_str);
	for (i = 0; i < num_subprograms; i++) {
		g_printerr ("%*s  nmtui %s [%s]\n",
		            nmt_newt_text_width (usage_str), " ",
		            subprograms[i].name,
		            _(subprograms[i].arg));
	}
	exit (1);
}

typedef struct {
	NmtuiSubprogram subprogram;
	int argc;
	char **argv;
} NmtuiStartupData;

static void
toplevel_form_quit (NmtNewtForm *form,
                    gpointer     user_data)
{
       nmtui_quit ();
}

static gboolean
idle_run_subprogram (gpointer user_data)
{
	NmtuiStartupData *data = user_data;
	NmtNewtForm *form;

	form = data->subprogram (TRUE, data->argc, data->argv);
	if (form) {
		g_signal_connect (form, "quit", G_CALLBACK (toplevel_form_quit), NULL);
		nmt_newt_form_show (form);
		g_object_unref (form);
	} else
		nmtui_quit ();

	return FALSE;
}

gboolean sleep_on_startup = FALSE;
gboolean noinit = FALSE;

GOptionEntry entries[] = {
	{ "sleep", 's', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &sleep_on_startup,
	  "Sleep on startup", NULL },
	{ "noinit", 'n', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &noinit,
	  "Don't initialize newt", NULL },
	{ NULL }
};

int
main (int argc, char **argv)
{
	GOptionContext *opts;
	GError *error = NULL;
	NmtuiStartupData startup_data;
	const char *prgname;
	int i;

	setlocale (LC_ALL, "");
	bindtextdomain (GETTEXT_PACKAGE, NMLOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	opts = g_option_context_new (NULL);
	g_option_context_add_main_entries (opts, entries, NULL);

	if (!g_option_context_parse (opts, &argc, &argv, &error)) {
		g_printerr ("%s: %s: %s\n",
		            argv[0],
		            _("Could not parse arguments"),
		            error->message);
		exit (1);
	}
	g_option_context_free (opts);

	nm_editor_bindings_init ();

	nm_client = nm_client_new (NULL, &error);
	if (!nm_client) {
		g_printerr (_("Could not contact NetworkManager: %s.\n"), error->message);
		g_error_free (error);
		exit (1);
	}
	if (!nm_client_get_nm_running (nm_client)) {
		g_printerr ("%s\n", _("NetworkManager is not running."));
		exit (1);
	}

	if (sleep_on_startup)
		sleep (5);

	startup_data.subprogram = NULL;
	prgname = g_get_prgname ();
	if (g_str_has_prefix (prgname, "lt-"))
		prgname += 3;
	if (!strcmp (prgname, "nmtui")) {
		if (argc > 1) {
			for (i = 0; i < num_subprograms; i++) {
				if (!strcmp (argv[1], subprograms[i].name)) {
					argc--;
					argv[0] = (char *) subprograms[i].shortcut;
					memmove (&argv[1], &argv[2], argc * sizeof (char *));
					startup_data.subprogram = subprograms[i].func;
					break;
				}
			}
		} else
			startup_data.subprogram = nmtui_main;
	} else {
		for (i = 0; i < num_subprograms; i++) {
			if (!strcmp (prgname, subprograms[i].shortcut)) {
				startup_data.subprogram = subprograms[i].func;
				break;
			}
		}
	}
	if (!startup_data.subprogram)
		usage ();

	if (!noinit)
		nmt_newt_init ();

	startup_data.argc = argc;
	startup_data.argv = argv;
	g_idle_add (idle_run_subprogram, &startup_data);
	loop = g_main_loop_new (NULL, FALSE);
	g_main_loop_run (loop);
	g_main_loop_unref (loop);

	if (!noinit)
		nmt_newt_finished ();

	g_object_unref (nm_client);

	return 0;
}
